x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java

```diff
@@ -361,68 +367,95 @@ private void prewarmCache() {
+                                    assert bytesRead == cacheFile.partBytes(part);
+
+                                    logger.trace(
+                                        () -> new ParameterizedMessage(
+                                            "part [{}/{}] of [{}] warmed in [{}] ms",
+                                            part,
+                                            numberOfParts,
+                                            fileName,
+                                            TimeValue.timeValueNanos(statsCurrentTimeNanosSupplier.getAsLong() - startTimeInNanos).millis()
+                                        )
+                                    );
+                                } catch (Exception e) {
+                                    logger.trace(
+                                        () -> new ParameterizedMessage(
+                                            "failed to warm cache for [{}] part [{}/{}]",
+                                            fileName,
+                                            part,
+                                            numberOfParts
+                                        ),
+                                        e
+                                    );
+                                    if (e instanceof AlreadyClosedException
+                                        || (e.getCause() != null && e.getCause() instanceof AlreadyClosedException)) {
+                                        return; // don't rethrow in case of cache eviction or directory closing
+                                    }
+                                    throw e;
+                                } finally {
+                                    if (countDown.countDown()) {
+                                        IOUtils.closeWhileHandlingException(input);
+                                    }
                                 }
-                            }
-                        });
+                            });
+                        }
                     }
                 } catch (IOException e) {
-                    logger.trace(() -> new ParameterizedMessage("failed to warm cache for [{}]", fileName), e);
+                    logger.trace(() -> new ParameterizedMessage("unable to schedule cache prewarming for [{}]", fileName), e);
                 }
             }
+
+            // Start as many workers as fit into the searchable snapshot pool at once at the most
+            final int workers = Math.min(threadPool.info(SEARCHABLE_SNAPSHOTS_THREAD_POOL_NAME).getMax(), queue.size());
+            for (int i = 0; i < workers; ++i) {
+                executeNextWarmer(executor, queue);
+            }
         }
     }
 
+    private void executeNextWarmer(final Executor executor, final BlockingQueue<CheckedRunnable<Exception>> queue) {
+        executor.execute(new AbstractRunnable() {
```

Comment:
If scheduling fails here, how do we guarantee that IndexInput is closed?

Reply:
Arghl right... you saw nothing. I'll address that.

![neuralyzer](https://user-images.githubusercontent.com/642733/80361162-c7f2f000-8880-11ea-8a72-773f3a864e1d.jpg)